### DISCUSSION TITLE
Address Hooks are different between BO and FO

### DISCUSSION THREAD

**Discussion by thalassa-web** [PrestaShop/PrestaShop#33418] [Ideas]
#### Describe the bug
In **F.O.**, when I update a customer's address (in my account), the hooks actionObjectAddress[mode] are triggered.
In **B.O.**, when I update a customer's address, the hooks actionObjectCustomerAddress[mode] are triggered.
It's quite  disturbing...

#### Expected behavior
Using the same rule in B.O. and F.O..
Or maybe call both to avoid backward compatibility issue …

#### Steps to Reproduce

1. Make a simple module and listen `actionObjectAddressAddBefore` and `actionObjectCustomerAddressAddBefore`.
2. Just log some dummy text in hooks.
3. Connect to **Front office** with existing customer.
4. Add an address in my account.
5. The «`actionObjectAddressAddBefore`» hook will be called.

 --
1. Connect to the **Back Office**.
2. Go to the customers addresses list.
3. Add a new address.
4. The «`actionObjectCustomerAddressAddBefore`» hook will be called.

#### Additional information

* PrestaShop version: 1.7.6.8
* PHP version: 7.1

**Comment by hibatallahAouadni** [comment]
Hello @thalassa-web 

Could you please provide a test module to reproduce the issue :pray: 
Waiting for your feedback.

Thanks!

**Comment by thalassa-web** [comment]
Hi @hibatallahAouadni 

I made a dummy module which logs in a file named test.txt. This file is placed in the directory of the module.
It will be created after the first address creation (via the BO or the FO).
It only write the name of the hook called actionObjectAddressAddAfter and actionObjectCustomerAddressAddAfter.

The module is available here : https://github.com/thalassa-web/thalassaweb25709/releases/download/v1.0.0/thalassaweb25709.zip

**Comment by khouloudbelguith** [comment]
Hi @thalassa-web,

Thanks for your feedback.
I did not manage to reproduce the issue with PS1777.
I attached a screen record
https://drive.google.com/file/d/1vkWHq1sS-GQYkAsVdmlpkHQUVmh-6uvp/view?usp=sharing

Thanks to check and feedback.

**Comment by urecheatu007** [comment]
The issue here is that in front we have Address and in backend CustomerAddress for the same Address class/object.

**Comment by thalassa-web** [comment]
@urecheatu007 I already know that. But it's quite weird. If someone change a customer address and I want to make some treatments  after the modification, the module has to listen two differents hooks to do the same thing.
And with the Address hook, the module has to check if it's a customer's one …
But apparently in the last version the problem does not exist anymore …

**Comment by matks** [comment]
It's probably a good thing that hooks for BO and FO are different. The two environments are highly different, and what a module can do in BO context or FO context varies a lot.

Having 2 hooks to handle the 2 updates seems a good idea to me, it allows modules interested only in FO updates to hook only on FO, and modules interested only in BO updates to hook only on BO. It's more modular.

BO and FO can almost be considered as 2 separate applications 😉  they are very different in structure and behavior. So 2 hooks for 2 different applications, seems OK to me.

**Comment by thalassa-web** [comment]
@matks Ok why not, but in this case is there any difference between Customer hook modification in the FO and in the BO ? If I change my account info, the hook actionObjectCustomerUpdateAfter will be triggered. If someone on the BO change some info in my account, the same hook will be triggered …

If we say : the hook when I change an object is different between the BO and the FO it has to be true for all hooks.
And I think that is not the case.
Furthermore, the module could do something different depending on the context (maybe if the curent controller is an Admin one, we are in a BO context …)

My point was about the consistency. I think (maybe I'm wrong) the difference between BO and FO for the actionObject hooks is only true for the address.

Secondly, if the two hooks depends on the BO or FO context, the name could contains this info (actionObjectAddAfterBO …). 
But with Address and CustomerAddress it's the kind of address which is important.
When I read the hook name, I think  : OK, actionObjectAddressAddAfter will be triggered for all kind of addresses. And with actionObjectCustomerAddressAddAfter, only triggered with a customer address (and not another kind …).

**Comment by urecheatu007** [comment]
Reading all the comments, it seems that the naming is indeed used to identify that those hooks are for the customer address, since the table is used for other types of addresses.
As a coincidence, today i was working on a module to add some fields to the address, and if you need to use those new fields somewhere in the FO (other than a form) you will need 2 overrides (in Address - just the fields and definition and Customer->getSimpleAddressSql()) , or just in Customer if you need those fields only in the checkout/addresses step (or you access them via $customer->addresses), at least in 1.7.7.1.
Some sort of a hook letting us defining new fields in any object would be a godsend - i am thinking about something like action<Object>Construct located somewhere in the constructor / before hidration so that we can define the new fields and add them to the $definition array  - this will cut the usage of the overrides by a large margin.

**Comment by kpodemski** [comment]
Since it is clearly an idea to improve the current state of things, I'm converting this issue into a discussion.
